src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxPermissionFilter.cs

```diff
@@ -87,38 +86,40 @@ private ITransactionPermissionContract.TxPermissions GetPermissions(Transaction
                 
                 try
                 {
-                    (txPermissions, shouldCache) = versionedContract.AllowedTxTypes(parentHeader, tx);
+                    (txPermissions, shouldCache, contractExists) = versionedContract.AllowedTxTypes(parentHeader, tx);
                 }
                 catch (AbiException e)
                 {
-                    if (_logger.IsError) _logger.Error($"Error calling tx permissions contract on {parentHeader.ToString(BlockHeader.Format.FullHashAndNumber)} for tx {tx.ToShortString()} {new StackTrace()}.", e);
+                    if (_logger.IsError) _logger.Error($"Error calling tx permissions contract on {parentHeader.ToString(BlockHeader.Format.FullHashAndNumber)} for tx {tx.ToShortString()}.", e);
                 }
             }
 
+            var result = (txPermissions, contractExists);
+            
             if (shouldCache)
             {
-                _cache.Permissions.Set(key, txPermissions);
+                _cache.Permissions.Set(key, result);
             }
 
-            return txPermissions;
+            return result;
         }
 
         private ITransactionPermissionContract? GetVersionedContract(BlockHeader blockHeader)
             => _contract.ResolveVersion(blockHeader);
 
-        private ITransactionPermissionContract.TxPermissions GetTxType(Transaction tx) =>
+        private ITransactionPermissionContract.TxPermissions GetTxType(Transaction tx, bool contractExists) =>
             tx.IsContractCreation
                 ? ITransactionPermissionContract.TxPermissions.Create
-                : (_stateProvider.GetCode(tx.To) ?? Array.Empty<byte>()).Length != 0
+                : contractExists
                     ? ITransactionPermissionContract.TxPermissions.Call
                     : ITransactionPermissionContract.TxPermissions.Basic;
         
         public class Cache
         {
             public const int MaxCacheSize = 4096;
             
-            internal ICache<(Keccak ParentHash, Address Sender), ITransactionPermissionContract.TxPermissions?> Permissions { get; } =
-                new LruCache<(Keccak ParentHash, Address Sender), ITransactionPermissionContract.TxPermissions?>(MaxCacheSize, "TxPermissions");
+            internal ICache<(Keccak ParentHash, Address Sender), (ITransactionPermissionContract.TxPermissions Permissions, bool ContractExists)> Permissions { get; } =
```

Comment:
can we use structs instead of Tuples here

Reply:
ValueTuples are structs, they are also fine in terms of performance of Equals and GetHashCode (I've just checked). As this is internal, well named and is not used in many places I think its fine and reduces boilerplate code.